apps/pwabuilder/src/script/components/share-card.ts

```diff
@@ -0,0 +1,291 @@
+    lengthOfDecodedData = decodedData.length,
+    u8array = new Uint8Array(lengthOfDecodedData);
+    while (lengthOfDecodedData--) {
+      u8array[lengthOfDecodedData] = decodedData.charCodeAt(lengthOfDecodedData);
+    }
+    return new File([u8array], filename, { type: mimeType });
+  };
+
+  shareFile(file: File, title: string, text: string) {
+    if (navigator.canShare && navigator.canShare({ files: [file] })) {
+      navigator
+        .share({
+          files: [file],
+          title,
+          text
+        })
+        .then(() => console.log("Share was successful."))
+        .catch((error) => console.log("Sharing failed", error));
+    } else {
+      console.log(`Your system doesn't support sharing files.`);
+    }
+  };
+
+  hideDialog(){
+    let dialog: any = this.shadowRoot!.querySelector(".dialog");
+    dialog!.hide();
+  }
+
+  handleRequestClose(e: Event){
+    if (this.preventClosing) {
+      e.preventDefault();
+    }
+  }
+
+  renderShareActionButtons(){
+    const file = this.dataURLtoFile(this.dataURL, `${this.siteName}_pwabuilder_score.png`);
+
+    if (!navigator.canShare || !navigator.canShare({files: [file]})) {
+      return;
+    } else {
+      this.actionButtons = html`
+        <div>
+          <button type="button" id="share-button" class="standard-button secondary" @click=${() => this.htmlToImage('share')}><img class="actions-icons" src="/assets/share_icon.svg" alt="Share image button icon"/></button>  
+          <span class="standard-button-label">Share</span>
+        </div>                          
+      `
+    }
+    return;
+  }
+
+  render() {
+    return html`
+      <sl-dialog class="dialog" @sl-show=${() => this.setup()} @sl-hide=${() => this.hideDialog()} noHeader>
+        <div class="share-modal-header">Share your PWA score with the community!</div>
+        <div id="frame-wrapper">
+          <div id="frame-content">
+            <div id="canvas-holder">
+              <canvas id="myCanvas" ${ref(this.shareCanvas)}>
+                Your browser does not support the canvas element.
+              </canvas>
+            </div>
+            <div id="share-actions">
+              <div>
+                <button type="button" id="copy-button" class="standard-button secondary"  @click=${() => this.copyImage()}><img class="actions-icons" src="/assets/copy-icon-standard-color.svg" alt="Download image button icon"/></button>
+                <span id="copy-button-label" class="standard-button-label">Copy</span>
+              </div>
+              <div>
+                <button type="button" id="download-button" class="standard-button secondary" @click=${() => this.htmlToImage('download')}><img class="actions-icons" src="/assets/download-icon-standard-color.svg" alt="Download image button icon"/></button>
+                <span id="download-button-label" class="standard-button-label">Download</span>
+              </div>
+              ${this.actionButtons}                  
```

Comment:
Is this will work instead of this.actionButtons?
```
${navigator.canShare && navigator.canShare({files: [file]})? html `<div>
          <button type="button" id="share-button" class="standard-button secondary" @click=${() => this.htmlToImage('share')}><img class="actions-icons" src="/assets/share_icon.svg" alt="Share image button icon"/></button>  
          <span class="standard-button-label">Share</span>
        </div>   `: ''} 
```
This one I guess you can put to setup?
```const file = this.dataURLtoFile(this.dataURL, `${this.siteName}_pwabuilder_score.png`);```

Reply:
It has to stay this way because canshare needs to be determined in setup not in render function or it will read as null